src/nix-eval-jobs.cc

```diff
@@ -384,16 +381,75 @@ int main(int argc, char * * argv)
             }
         };
 
-        EvalState initialState(myArgs.searchPath, openStore());
-        Bindings & autoArgs = *myArgs.getAutoArgs(initialState);
+        /* Collect initial attributes to evaluate */
+        {
+            AutoCloseFD from, to;
+            Pipe toPipe, fromPipe;
+            toPipe.create();
+            fromPipe.create();
+            Pid p = startProcess(
+                [&,
+                 to{std::make_shared<AutoCloseFD>(std::move(fromPipe.writeSide))},
+                 from{std::make_shared<AutoCloseFD>(std::move(toPipe.readSide))}
+                ]()
+                {
+                    debug("created initial attribute collection process %d", getpid());
+
+                    try {
+                        EvalState state(myArgs.searchPath, openStore());
+                        Bindings & autoArgs = *myArgs.getAutoArgs(state);
+
+                        auto vRoot = releaseExprTopLevelValue(state, autoArgs);
+
+                        if (vRoot->type != tAttrs)
+                            throw TypeError("top level value is a '%s', expected an attribute set", showType(*vRoot));
+
+                        nlohmann::json reply;
+
+                        try {
+                            std::vector<std::string> attrs;
+                            for (auto & a : vRoot->attrs->lexicographicOrder())
+                                attrs.push_back(a->name);
+
+
+                            reply["attrs"] = attrs;
+
+                            writeLine(to->get(), reply.dump());
 
-        auto topLevelValue = releaseExprTopLevelValue(initialState, autoArgs);
+                        } catch (EvalError & e) {
+                            reply["error"] = filterANSIEscapes(e.msg(), true);
+                            printError(e.msg());
 
-        if (topLevelValue->type == tAttrs) {
-          auto state(state_.lock());
-          for (auto & a : topLevelValue->attrs->lexicographicOrder()) {
-            state->todo.insert(a->name);
-          }
+                        }
+                        writeLine(to->get(), reply.dump());
```

Comment:
Is this second `writeLine` necessary?  It looks like the same logic as on line 417

Reply:
Ah! Good catch. I think this inner `try/catch` is also redundant since these `EvalError`s like `TypeError` are subclasses of `Error`.

Reply:
I think `cd90c56` should fix it. Much better!